core/src/main/java/feign/ReflectiveFeign.java

```diff
@@ -160,21 +167,16 @@ public String toString() {
     protected final MethodMetadata metadata;
     private final Map<Integer, Expander> indexToExpander = new LinkedHashMap<Integer, Expander>();
 
-    private BuildTemplateByResolvingArgs(MethodMetadata metadata) {
+    private BuildTemplateByResolvingArgs(MethodMetadata metadata, Factory paramExpanderFactory) {
       this.metadata = metadata;
+      checkNotNull(paramExpanderFactory, "paramExpanderFactory");
       if (metadata.indexToExpanderClass().isEmpty()) {
         return;
       }
       for (Entry<Integer, Class<? extends Expander>> indexToExpanderClass : metadata
```

Comment:
great point on back compat..

In this case, I'd make this simpler.. make MethodMetadata.indexToExpander
nullable. Document it as if unset, this will be instantiated via
newInstance.

If MethodMetadata.indexToExpander is set (by a Contract) by the time
ReflectiveFeign is trying to assign indexToExpander, use it. Otherwise,
default to the existing logic.

Make sense?